packages/js/src/cache/notifications-cache.ts

```diff
@@ -0,0 +1,253 @@
+import { InMemoryCache } from './in-memory-cache';
+import type { Cache } from './types';
+import type { ListNotificationsArgs, ListNotificationsResponse, Notification } from '../notifications';
+import { NotificationEvents, NotificationsEvents, NovuEventEmitter } from '../event-emitter';
+import type { NotificationFilter } from '../types';
+import { areTagsEqual, isSameFilter } from '../utils/notification-utils';
+
+const getCacheKey = ({ tags, read, archived, limit, offset, after }: ListNotificationsArgs): string => {
+  return JSON.stringify({ tags, read, archived, limit, offset, after });
+};
+
+const getFilterKey = ({ tags, read, archived }: Pick<ListNotificationsArgs, 'tags' | 'read' | 'archived'>): string => {
+  return JSON.stringify({ tags, read, archived });
+};
+
+const getFilter = (key: string): NotificationFilter => {
+  return JSON.parse(key);
+};
+
+// these events should update the notification in the cache
+const notificationEvents: NotificationEvents[] = [
+  'notification.read.pending',
+  'notification.read.resolved',
+  'notification.unread.pending',
+  'notification.unread.resolved',
+  'notification.complete_action.pending',
+  'notification.complete_action.resolved',
+  'notification.revert_action.pending',
+  'notification.revert_action.resolved',
+];
+
+// these events should remove the notification from the cache
+const notificationEventsThatRemove: NotificationEvents[] = [
+  'notification.archive.pending',
+  'notification.unarchive.pending',
+];
+
+// these events should update the notifications in the cache
+const notificationsEvents: NotificationsEvents[] = [
+  'notifications.read_all.pending',
+  'notifications.read_all.resolved',
+];
+
+// these events should remove the notifications from the cache
+const notificationsEventsThatRemove: NotificationsEvents[] = [
+  'notifications.archive_all.pending',
+  'notifications.archive_all_read.pending',
+];
+
+export class NotificationsCache {
+  #emitter: NovuEventEmitter;
+  /**
+   * The key is the stringified notifications filter, the values are the paginated notifications.
+   */
+  #cache: Cache<ListNotificationsResponse>;
+
+  constructor() {
+    this.#emitter = NovuEventEmitter.getInstance();
+    notificationEvents.forEach((event) => {
```

Comment:
Instead of splitting events per resource and making the cache know about the notification types, group them by cache functionality. For example, updateEvents and clearanceEvents. That way, you will have two arrays to iterate and add listeners on.

Moreover, pass these events to the cosntructors so that the cache class is unaware of the events types in the first place. Pure dependency injection ;)

Reply:
The thing is that the events are tied to a callback interface implementation, so DI is not suited well here.